### PR TITLE
scripts(termux_download): handle fail download midway and retry

### LINKED ISSUE #20036: [Enhancement]: applying speed limits to `termux_download`
Can we please make termux_download retry if speed falls to 0 for 30 seconds or longer? 
Sometimes connection stucks for 10 minutes or more so updater-script is simply waiting for new data.
According to `curl`'s man we should use `--speed-limit` and `--speed-time` options.
https://github.com/termux/termux-packages/blob/5cf7e74e6d88a8a029e9e06825c33c6b32cedd53/scripts/build/termux_download.sh#L21

### PR DESCRIPTION
**Pull request by truboxl**
Also set minimum download speed threshold

Close #20036

### CODE REVIEW

File: scripts/build/termux_download.sh [hunk 1]
```diff
@@ -1,53 +1,68 @@
...
-	if curl "${CURL_OPTIONS[@]}" --output "$TMPFILE" "$URL"; then
-		local ACTUAL_CHECKSUM
-		ACTUAL_CHECKSUM=$(sha256sum "$TMPFILE" | cut -f 1 -d ' ')
-		if [ "$CHECKSUM" != "SKIP_CHECKSUM" ]; then
-			if [ "$CHECKSUM" != "$ACTUAL_CHECKSUM" ]; then
-				>&2 printf "Wrong checksum for %s\nExpected: %s\nActual:   %s\n" \
-					   "$URL" "$CHECKSUM" "$ACTUAL_CHECKSUM"
-				return 1
+	if ! curl "${CURL_OPTIONS[@]}" --output "$TMPFILE" "$URL"; then
+		local e=1
```
Comment (Biswa96) [inline scripts/build/termux_download.sh:35]: "Well-chosen identifiers make it significantly easier for developers" - https://en.wikipedia.org/wiki/Naming_convention_(programming)

Comment (TomJo2000) [reply scripts/build/termux_download.sh:35]: Yep, `err` or `error` would probably be preferable.

File: scripts/build/termux_download.sh [hunk 2]
```diff
@@ -1,53 +1,68 @@
...
+		local retry=2
+		local try
+		for try in $(seq "${retry}"); do
```
Comment (TomJo2000) [inline scripts/build/termux_download.sh:38]:
```suggestion
		for (( try=1; try <= retry; try++ )); do
```

File: scripts/build/termux_download.sh [hunk 3]
```diff
@@ -1,53 +1,68 @@
...
 
 	local TMPFILE
 	local -a CURL_OPTIONS=(
 		--fail
-		--retry 20
+		--retry 10
 		--retry-connrefused
-		--retry-delay 30
+		--retry-delay 5
+		--speed-limit 10000
```
Comment (TomJo2000) [inline scripts/build/termux_download.sh:24]: To clarify, this is a lower bound.
Not an upper bound?

File: scripts/build/termux_download.sh [hunk 4]
```diff
@@ -1,53 +1,68 @@
...
-				>&2 printf "Wrong checksum for %s\nExpected: %s\nActual:   %s\n" \
-					   "$URL" "$CHECKSUM" "$ACTUAL_CHECKSUM"
-				return 1
+	if ! curl "${CURL_OPTIONS[@]}" --output "$TMPFILE" "$URL"; then
+		local e=1
+		local retry=2
+		local try
+		for try in $(seq "${retry}"); do
+			echo "Retrying #${try} download ${URL}"
+			sleep 1
```
Comment (agnostic-apollo) [inline scripts/build/termux_download.sh:40]: Would be better to increase the sleep time to `60s` or `30s` at least. Previously, total retry time was `570s`/`9.5min`, where there was a wait of `30s` between each of the `20` tries. Now, the total retry time will be `45s`, where there is a wait of `5s` between each of the `10` tries, and then another retry period of `45s` after `1s`, so total `91s`/`1min31s`. It's very likely that sometimes servers will become inactive for longer than that. If we increase sleep time to `60s`, then total retry period would be `2min31s`, its still `~1/4` of the previous `10min`, so much lower with lower chance of build failure. The above calculations assume that the last retry wait doesn't happen as that would be after the last retry, and some time will be spent by `curl` on each retry until it fails/times out, so actual execution time would be longer.

File: scripts/build/termux_download.sh (hunk 3 above)
Comment (agnostic-apollo) [reply scripts/build/termux_download.sh:24]: Yes, it is a lower bound. And would be better to change it to `1000` bytes, `curl` uses `1` byte by default if `--speed-time` is passed without `--speed-limit`. I am pretty sure `1000` bytes/`1KB` could happen for some very slow source repos and should still indicate a working download connection.

https://manpages.debian.org/testing/curl/curl.1.en.html#Y,

File: scripts/build/termux_download.sh [hunk 5]
```diff
@@ -1,53 +1,69 @@
...
-	return 1
+	local ACTUAL_CHECKSUM
+	ACTUAL_CHECKSUM=$(sha256sum "$TMPFILE" | cut -d' ' -f1)
+	if [[ -z "$CHECKSUM" ]]; then
+		printf "WARNING: No checksum check for %s:\nActual: %s\n" \
+			"$URL" "$ACTUAL_CHECKSUM"
+	elif [[ "$CHECKSUM" == "SKIP_CHECKSUM" ]]; then
+		:
+	elif [[ "$CHECKSUM" != "$ACTUAL_CHECKSUM" ]]; then
+		>&2 printf "Wrong checksum for %s\nExpected: %s\nActual:   %s\n" \
```
Comment (agnostic-apollo) [inline scripts/build/termux_download.sh:61]: Redirection should be after `printf` like other other `echo` errors.

**Comment by agnostic-apollo** [comment]
The `--retry` argument has been reverted back to `5`. I guess we can use that for now and see. Pull looks fine otherwise. It should be merged as I need to rebase against the changes for my pull, I can squash and "rebase and merge" if you want me to do it.

**Comment by Dgutfhjugugfftzuzg** [comment]
Screenshot_٢٠٢٤٠٩٠٨_٠٠٣٦١٨

**Comment by truboxl** [comment]
@agnostic-apollo can squash and merge

**Comment by agnostic-apollo** [comment]
Done. Thanks.

**Comment by zx274** [comment]
welkam_~~ yj~~_

**Pull request merged**